### DISCUSSION TITLE
Correct usage of CASE?

### DISCUSSION THREAD

**Discussion by ianjosephwilson** [sqlalchemy/sqlalchemy#6785] [Usage Questions]
Trying to bring my code base into the 2020s and going through the warnings and saw a warning for my case statement.

```python
        whens = {}
        for size in ('1gal', '5gal', '15gal'):
            whens.update({
                f'{size}-fr': f'{size}-inventory',
                f'{size}-pr': f'{size}-production',
                f'{size}-nr': f'{size}-production'
            })
        category = case(whens=whens, value=items_t.c.short_code, else_=None).label('category')
```

Using a dict still seems ok in the docs.

Seems that the "Was a list but is now position args" warning should trigger in the else statement instead of this if statement??:

https://github.com/sqlalchemy/sqlalchemy/blob/a25d6ead9cf6d2d51ae67e53eac50f642a0bb432/lib/sqlalchemy/sql/elements.py#L2920

**Comment by CaselIT** [comment]
Hi,

The documentation is outdated, the attribute description is correct: https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.case.params.*whens

It should work by removing the whens keyword from your example:
```py
category = case(whens, value=items_t.c.short_code, else_=None).label('category')
```

**Reply by CaselIT** [reply]
Updated here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2959

**Reply by ianjosephwilson** [reply]
It seems that the `if` block emits the warning when `whens` is a kw and so does the `else` block via `coercions._expression_collection_was_a_list` because `*whens` is a single element tuple containing the dict.  Not sure but it appears the warning is unsquelchable.

```python
def _expression_collection_was_a_list(attrname, fnname, args):
    if args and isinstance(args[0], (list, set, dict)) and len(args) == 1:
        util.warn_deprecated_20(
            'The "%s" argument to %s() is now passed as a series of '
            "positional "
            "elements, rather than as a list. " % (attrname, fnname)
        )
        return args[0]
    else:
        return args

```
```python

        #...
        if "whens" in kw:
            util.warn_deprecated_20(
                'The "whens" argument to case() is now passed as a series of '
                "positional "
                "elements, rather than as a list. "
            )
            whens = kw.pop("whens")
        else:
            whens = coercions._expression_collection_was_a_list(
                "whens", "case", whens
            )
```

**Reply by ianjosephwilson** [reply]
If passing a dict as a single positional argument is the expected usage then I think `coercions._expression_collection_was_a_list` will need to be avoided when `len(whens) == 1 and type(whens[0]) == dict`.

**Reply by CaselIT** [reply]
Yes, that seems to be a bug. Opening an issue

**Reply by CaselIT** [reply]
see  #6786
